packages/components/src/color-picker/test/index.tsx

```diff
@@ -342,4 +345,89 @@ describe( 'ColorPicker', () => {
 			} );
 		} );
 	} );
+
+	describe.each( [
+		[ 'hsl', 'HSL' ],
+		[ 'rgb', 'RGB' ],
+	] )( 'Alpha-enabled %s format', ( format, formatLabel ) => {
+		it( `should update alpha correctly when ${ formatLabel } format is selected`, async () => {
+			const user = userEvent.setup();
+			const onChange = jest.fn();
+
+			render(
+				<ControlledColorPicker
+					onChange={ onChange }
+					enableAlpha
+					initialColor="#ffffff80"
+				/>
+			);
+
+			const formatSelector = screen.getByRole( 'combobox' );
+			expect( formatSelector ).toBeVisible();
+			await user.selectOptions( formatSelector, format );
+
+			const alphaInput = screen.getByRole( 'spinbutton', {
+				name: 'Alpha',
+			} );
+			expect( alphaInput ).toBeVisible();
+
+			const alphaSliders = screen.getAllByRole( 'slider', {
+				name: 'Alpha',
+			} );
+
+			expect( alphaSliders ).toHaveLength( 2 );
+
+			// Choose the second slider which is the actual slider of type: input[type="range"]
+			const alphaSlider = alphaSliders.at( -1 )!;
+
+			expect( alphaSlider ).toHaveValue( '50' );
+			expect( alphaInput ).toHaveValue( 50 );
+
+			expect( onChange ).not.toHaveBeenCalledWith();
+
+			// Test pattern 1: Update the slider
+			fireEvent.change( alphaSlider, {
+				target: { value: 75 },
+			} );
+
+			await waitFor( () => {
+				expect( onChange ).toHaveBeenCalledTimes( 1 );
+			} );
+			await waitFor( () => {
+				expect( onChange ).toHaveBeenCalledWith( '#ffffffbf' );
+			} );
+
+			await waitFor( () => {
+				expect( alphaInput ).toHaveValue( 75 );
+			} );
+			await waitFor( () => {
+				expect( alphaSlider ).toHaveValue( '75' );
+			} );
+
+			onChange.mockClear();
+
+			// Test pattern 2: Update the alphaInput
+			await user.clear( alphaInput );
+			expect( onChange ).toHaveBeenCalledTimes( 1 );
+
+			// Initially type 7 in the alpha input, we expect it to be called with #ffffff12
+			await user.type( alphaInput, '7' );
+
+			expect( onChange ).toHaveBeenCalledTimes( 2 );
+			await waitFor( () => {
+				expect( onChange ).toHaveBeenCalledWith( '#ffffff12' );
+			} );
+
+			// Now with 75% opacity we expect it to be called with #ffffffbf
+			await user.type( alphaInput, '5' );
+
+			expect( onChange ).toHaveBeenCalledTimes( 3 );
+			await waitFor( () => {
+				expect( onChange ).toHaveBeenCalledWith( '#ffffffbf' );
+			} );
```

Comment:
`user.type` should already simulate distinct keypresses for each character of the string ([docs](https://testing-library.com/docs/user-event/utility/#type))

I wonder if we could change the test to something like:

```suggestion
			await user.type( alphaInput, '75' );
			
			await waitFor( () => {
				// Called twice, once per key stroke (`7` and `5`)
				expect( onChange ).toHaveBeenCalledTimes( 3 );
			} );

			// Called first with `5` and then `75` % opacity
			expect( onChange ).toHaveBeenNthCalledWith(2, '#ffffff12');
			expect( onChange ).toHaveBeenNthCalledWith(3, '#ffffffbf');
```

Reply:
Thanks for the feedback @ciampo .

Well, I tried this earlier as well. 

But when actually implementing this, I notice a really weird behavior. If we type a 3-digit number, let's say 100, then the alphaInput would have all 3 digits (i.e., 100). But if we type a 2-digit number like 75 or any other, then only the first digit is received.

for a 2-digit number: user.type( alphaInput, '75' ) triggers onChange 1 time ie only for 7


The problem is weird. I tried debugging for an hour or so, but I don't know why this happens. The same user.type has been used in numerous places in the codebase; there it works as expected. I'm not sure what could be causing this problem.

I tried multiple different approaches like adding a timeout to give enough time for the type event to complete, added `waitFor(() => {})`, etc. None of them worked correctly.


This snippet works well, but its based on `fireEvent` and not user.type event. 

```js
fireEvent.change( alphaInput, { target: { value: '75' } } );

await waitFor( () => {
expect( onChange ).toHaveBeenCalledTimes( 2 );
} );
expect( onChange ).toHaveBeenNthCalledWith( 2, '#ffffffbf' );

expect( alphaSlider ).toHaveValue( '75' );

expect( alphaInput ).toHaveValue( 75 );
```

I would appreciate your feedback on this issue, and help if you have faced this kind of problem before.

Reply:
Currently I have added this code https://github.com/WordPress/gutenberg/pull/69203/commits/8a05ee167e4a5a3a4d5ae641f5ea06f547fa8252, it's as close as the above suggested snippet, barring the use of 2 `user.type()` for 7 and 5 individually.

Reply:
That's weird — it's clearly not working as expected (other folks seem to [have related issues too](https://github.com/testing-library/user-event/issues/1150)).

Let's switch to using `user.keyboard` instead of `user.type`, at least we get the expected behavior:

```ts
			// Initially type 7 in the alpha input, we expect it to be called with #ffffff12
			await user.keyboard( '7' );

			// Now with 75% opacity we expect it to be called with #ffffffbf
			await user.keyboard( '5' );
```

Reply:
Works like a charm thanks for the quick input 🙇